### PR TITLE
Abstract `ChannelManager` outbound payment logic 

### PR DESCRIPTION
**Pull request by valentinewallace**
Moves `ChannelManager` outbound payment logic into its own module, in preparation for payment retries. 

Just code moves mostly. I kept the serialization changes surgical but we could probably move more logic there too.

### CODE REVIEW

File: lightning/src/ln/mod.rs [hunk 1]
```diff
@@ -32,6 +32,7 @@ pub mod peer_handler;
 pub mod chan_utils;
 pub mod features;
 pub mod script;
+mod outbound_payment;
```
Comment (dunxen) [inline lightning/src/ln/mod.rs:35]: nit: Is it also worth ungrouping this?

**Review by dunxen — COMMENTED** [review]
The moves and renames LGTM!

File: lightning/src/ln/outbound_payment.rs [hunk 2]
```diff
@@ -248,6 +260,547 @@ pub enum PaymentSendFailure {
...
+						});
+				}
+			}
+		} else {
+			log_trace!(logger, "Received duplicative fulfill for HTLC with payment_preimage {}", log_bytes!(payment_preimage.0));
+		}
+		(payment_sent_ev, path_success_ev)
+	}
+
+	pub(super) fn finalize_claims(&self, mut sources: Vec<HTLCSource>) -> Vec<events::Event> {
```
Comment (jkczyz) [inline lightning/src/ln/outbound_payment.rs:606]: nit: Doesn't need to be `mut` if you drop `drain(..)`

**Review by dunxen — DISMISSED** [review]

File: lightning/src/ln/channelmanager.rs [hunk 3]
```diff
@@ -3722,90 +3478,9 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
...
-							#[cfg(test)]
-							error_data: onion_error_data
-						}
-					}
-				};
+				let (path_failure, full_failure_ev) = self.pending_outbound_payments.fail_htlc(source, payment_hash, onion_error, path, session_priv, payment_id, payment_params, self.probing_cookie_secret, &self.secp_ctx, &self.logger);
 				let mut pending_events = self.pending_events.lock().unwrap();
-				pending_events.push(path_failure);
+				if let Some(ev) = path_failure { pending_events.push(ev); }
 				if let Some(ev) = full_failure_ev { pending_events.push(ev); }
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:3484]: Yeah, it does in some cases avoid holding the locks at the same time, but seems cleaner this way. Will switch.

**Comment by valentinewallace** [comment]
Lmk when I can squash, @jkczyz

File: lightning/src/ln/outbound_payment.rs [hunk 4]
```diff
@@ -271,9 +271,9 @@ impl OutboundPayments {
 		}
 	}
 
-	pub(super) fn send_payment<K: Deref, F>(&self, route: &Route, payment_hash: PaymentHash,
-		payment_secret: &Option<PaymentSecret>, payment_id: PaymentId, keys_manager: &K,
-		best_block_height: u32, send_payment_along_path: F)
+	pub(super) fn send_payment<K: Deref, F>(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>,
+		payment_id: PaymentId, keys_manager: &K, best_block_height: u32, send_payment_along_path: F)
```
Comment (jkczyz) [inline lightning/src/ln/outbound_payment.rs:276]: I think rustfmt would move the `)` to the same line as `->`.

**Review by jkczyz — COMMENTED** [review]
Feel free to squash

**Review by dunxen — APPROVED** [review]
The commits were pretty easy to follow! Cleans up `ChannelManager` a load.

File: lightning/src/ln/channelmanager.rs [hunk 5]
```diff
@@ -2580,142 +2368,27 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
...
-					assert!(payment.insert(*session_priv_bytes, path));
-				}
-
-				Ok(onion_session_privs)
-			},
-		}
+	#[cfg(test)]
+	pub(crate) fn test_add_new_pending_payment(&self, payment_hash: PaymentHash, payment_secret: Option<PaymentSecret>, payment_id: PaymentId, route: &Route) -> Result<Vec<[u8; 32]>, PaymentSendFailure> {
+		let best_block_height = self.best_block.read().unwrap().height();
+		self.pending_outbound_payments.add_new_pending_payment(payment_hash, payment_secret, payment_id, route, &self.keys_manager, best_block_height)
```
[the remark this answers is not in the document]
Comment (dunxen) [reply lightning/src/ln/channelmanager.rs:2389]: Yeah, it's definitely safer. Good for a follow-up.

File: lightning/src/ln/outbound_payment.rs [hunk 6]
```diff
@@ -0,0 +1,830 @@
...
+			Some(RouteParameters {
+				payment_params: payment_params_data.clone(),
+				final_value_msat: path_last_hop.fee_msat,
+				final_cltv_expiry_delta: path_last_hop.cltv_expiry_delta,
+			})
+		} else { None };
+		log_trace!(logger, "Failing outbound payment HTLC with payment_hash {}", log_bytes!(payment_hash.0));
+
+		let path_failure = {
+	#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (dunxen) [reply lightning/src/ln/outbound_payment.rs:715]: I've personally seen more code where it has the same indentation as the line itself (and not ifdef style). Don't think there's a standard though.

[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/ln/outbound_payment.rs:715]: Likwise, I've mostly noticed them indented.

File: lightning/src/ln/outbound_payment.rs [hunk 7]
```diff
@@ -433,8 +435,8 @@ impl OutboundPayments {
 		}
 	}
 
-	fn send_payment_internal<K: Deref, F>
-		(&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>,
+	fn send_payment_internal<K: Deref, F>(
```
Comment (jkczyz) [inline lightning/src/ln/outbound_payment.rs:438]: nit: I'd also move the close `)` to the next line here and all the following methods as you did in the methods before this (ec55730959103b71417d23d3a77011ca7d7d4ee5).

**Review by jkczyz — APPROVED** [review]

File: lightning/src/ln/outbound_payment.rs [hunk 8]
```diff
@@ -0,0 +1,830 @@
...
+	where L::Target: Logger {
+		let mut session_priv_bytes = [0; 32];
+		session_priv_bytes.copy_from_slice(&session_priv[..]);
+		let mut outbounds = self.pending_outbound_payments.lock().unwrap();
+		let mut all_paths_failed = false;
+		let mut full_failure_ev = None;
+		if let hash_map::Entry::Occupied(mut payment) = outbounds.entry(*payment_id) {
+			if !payment.get_mut().remove(&session_priv_bytes, Some(&path)) {
+				log_trace!(logger, "Received duplicative fail for HTLC with payment_hash {}", log_bytes!(payment_hash.0));
+				return
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply lightning/src/ln/outbound_payment.rs:684]: Oops, that mistakenly introduced when parameterizing the method by `pending_events`. I'll fix if Jeff has feedback that needs to go in this PR.

**Comment by valentinewallace** [comment]
Will address the latest feedback in a new PR shortly!

**Pull request merged**